mantidproject/mantid
# Add an Autoreduction configurator in SNAPReduce

Companion to PR #31771; Implements [ENG55](https://code.ornl.gov/sns-hfir-scse/diffraction/engineering-diffraction/engineering-diffraction/-/issues/55)

**Description of work.**
- [x] Add property for autoreduction configurator
- [x] Validator for selected algorithm properties
- [x] Save dictionary of selected algorithm properties to file in JSON format
- [x] Update algorithm documentation
- [x] Workbench test

**To test:**  
For this test to run, you will need to do the following in Mantid's workbench

- In `Settings`, set your `Facility` to `SNS`
- In `Manage User Directories`, enable the `Search Data Archive` to the default facility

In the workbench, run:

```python
from mantid.simpleapi import SNAPReduce

SNAPReduce(RunNumbers='51284', Masking='Horizontal', Binning='0.7,-0.004,5.5',
           Normalization='Extracted from Data', PeakClippingWindowSize=7,
           SmoothingRange=5, GroupDetectorsBy='2_4 Grouping', SaveData=True,
           OutputDirectory='/tmp',
           EnableConfigurator=True, ConfigSaveDir='/tmp')
```

Now in directory `/tmp/` search for files with extension `.json`. The most recent of those files should have the following contents:

```bash
{
    "Binning": [
        0.7,
        -0.004,
        5.5
    ],
    "GroupDetectorsBy": "2_4 Grouping",
    "Masking": "Horizontal",
    "Normalization": "Extracted from Data",
    "PeakClippingWindowSize": 7,
    "SaveData": true,
    "SmoothingRange": 5
}
```

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
